datumaro/plugins/coco_format/converter.py

```diff
@@ -658,17 +658,27 @@ def apply(self):
             if CocoTask.panoptic in task_converters:
                 self._make_segmentation_dir(subset_name)
 
-            for item in subset:
-                if self._save_images:
-                    if item.has_image:
-                        self._save_image(item, subdir=osp.join(self._images_dir,
-                            '' if self._merge_images else subset_name))
-                    else:
-                        log.debug("Item '%s' has no image info", item.id)
-                for task_conv in task_converters.values():
-                    task_conv.save_image_info(item,
-                        self._make_image_filename(item))
-                    task_conv.save_annotations(item)
+            for item in self._with_progress(subset,
+                    desc=f'Exporting {subset_name}'):
+                try:
+                    if self._save_images:
+                        if item.has_image:
+                            self._save_image(item, subdir=osp.join(
+                                self._images_dir,
+                                '' if self._merge_images else subset_name))
+                        else:
+                            log.debug("Item '%s' has no image info", item.id)
+
+                    for task_conv in task_converters.values():
+                        try:
+                            task_conv.save_image_info(item,
+                                self._make_image_filename(item))
+                            task_conv.save_annotations(item)
+                        except Exception as e:
+                            self._report_annotation_error(e,
+                                item_id=(item.id, item.subset))
+                except Exception as e:
+                    self._report_item_error(e, item_id=(item.id, item.subset))
```

Comment:
This pattern for exception handling seems quite common; perhaps it would be useful to factor it out as a context manager? E.g.:

```python
with self._item_error_handler(item_id=(item.id, item.subset)):
    if self._save_images: ...
        ...
```

Reply:
Unfortunately it has 2 major disadvantages:

- It will pessimize annotation import performance for no reason (we expect 1M calls):

```
def f1():
    try:
        x = 1
    except Exception as e:
        raise

def f2():
    x = 1

@contextlib.contextmanager
def cm():
    try:
        yield
    finally:
        pass

def f3():
    with cm():
        x = 1

print(timeit.timeit('f1()', globals=globals(), number=1000000))
print(timeit.timeit('f2()', globals=globals(), number=1000000))
print(timeit.timeit('f3()', globals=globals(), number=1000000))

>>>
0.06721735099563375
0.05293095999513753
1.3045302460086532
```

- It will only cover simplest cases like this. In fact, it is expected that with this functionality we'll be able to take different actions (more than just skipping the problem) depending on eg. user choice or predefined config. Looks quite optimistic IMO, but anyway. If we'll support multiple actions, the `except` clause looks more appropriate choice. I think such change would be a premature optimization.

Reply:
Okay, fair enough. Although FWIW, a 1 second slowdown for a _million_ items doesn't seem bad at all.

Reply:
It can be useful it general, because we may need to do some cleanup on skipped export errors, and we also may want to always close the reporter.